# ** Phần A 10 câu hỏi
1. Liệu hành khách ở khoang vip 1 tỉ lệ sống cao hơn không?( Tôi dự đoán kết quả sẽ là không)
2. Tỉ lệ sống sót giữa hai giới tính? (Tôi dự đoán kết quả sẽ là Nữ sẽ cao hơn)
3. Tỉ lệ trẻ em sống sót? ( Tôi dự đoán kết quả sẽ là <20%)
4. Những người đi cùng anh chị em hoặc vợ/chồng có tỉ lệ sống sót cao hơn người đi một mình không? (Dự đoán: Có, vì họ hỗ trợ nhau)
5. Tỉ lệ sống sót giữa các hạng có chênh lệch cao không? (Dự đoán: chênh lệch không cao)
6. Có cảng nào (embarked) tỉ lệ sống sót cao hơn còn lại không? ( Dự đoán : Có)
7. Cột nào sẽ có dữ liệu thiếu nhiều nhất? ( Dự đoán: cột tuổi)
8. Có phải người càng lớn tuổi tỉ lệ sống sót càng thấp? ( Dự đoán : đúng, vì không đủ sức chịu đựng)
9. Những hành khách mua vé giá cao (Top 5% hoặc 10%) có tỉ lệ sống sót gần như tuyệt đối không? (Dự đoán: Có, vì họ được ưu tiên cứu hộ trước) .
10. Những khoang nào có tỉ lệ sống cao nhất? (Dự đoán: các khoang phía mũi tàu)

# Phần B


In [33]:
import seaborn as sns
import pandas as pd

# Load dataset Titanic từ thư viện seaborn
df = sns.load_dataset('titanic')

# Hiển thị 5 dòng đầu
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


* Kết quả của hàm head() cho phép ta nhìn sơ lược về cấu trúc các cột và các giá trị ban đầu của tập dữ liệu. Qua đây, ta có thể thấy các biến quan trọng như survived

In [34]:
# Hiển thị 5 dòng cuối
df.tail()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
886,0,2,male,27.0,0,0,13.00,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.00,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.45,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.00,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.0,0,0,7.75,Q,Third,man,True,NaN,Queenstown,no,True


* Hàm tail() giúp xác nhận tổng số lượng dòng dữ liệu thực tế và kiểm tra xem dữ liệu ở phần cuối có bị lỗi hay lệch định dạng không.

In [35]:
# Kiểm tra kích thước tập dữ liệu
df.shape

(891, 15)

* Kết quả shape cho biết bộ dữ liệu Titanic gồm có 891 dòng và 15 cột. Điều này giúp ta hình dung được quy mô của dữ liệu để lựa chọn các kỹ thuật xử lý phù hợp sau này.

In [36]:
# Kiểm tra kiểu dữ liệu và giá trị thiếu
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    str     
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    str     
 8   class        891 non-null    category
 9   who          891 non-null    str     
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    str     
 13  alive        891 non-null    str     
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), str(5)
memory usage: 100.4 KB


* Hàm info() cung cấp cái nhìn chi tiết về kiểu dữ liệu (như float, int, object) và số lượng giá trị không trống (non-null) của từng cột. Ta có thể quan sát thấy các cột như age, embarked, và đặc biệt là deck có khá nhiều giá trị thiếu cần được xử lý.

In [37]:
# Thống kê mô tả cho các cột dữ liệu số
df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


* Hàm describe() cung cấp các chỉ số thống kê quan trọng như trung bình (mean), độ lệch chuẩn, giá trị lớn nhất và nhỏ nhất của các cột số. Ví dụ, nhìn vào cột fare, ta thấy có sự chênh lệch rất lớn giữa giá vé trung bình và giá vé cao nhất, cho thấy có sự phân hóa giàu nghèo rõ rệt trên tàu

In [38]:
# Tính số lượng giá trị thiếu
missing_count = df.isnull().sum()

# Tính phần trăm giá trị thiếu
missing_pct = (df.isnull().sum() / len(df)) * 100

# Gộp lại thành một bảng để dễ quan sát
missing_df = pd.DataFrame({'Số lượng': missing_count, 'Phần trăm (%)': missing_pct})
print(missing_df[missing_df['Số lượng'] > 0])

             Số lượng  Phần trăm (%)
age               177      19.865320
embarked            2       0.224467
deck              688      77.216611
embark_town         2       0.224467


* Cột có vấn đề: Cột deck (hoặc cabin) có tỉ lệ thiếu cao nhất (hơn 77%), tiếp theo là cột age (khoảng 19%) và embarked thiếu rất ít.
* Tại sao: deck/cabin thiếu nhiều có thể do hành khách hạng phổ thông không có vị trí phòng cố định hoặc dữ liệu không được ghi chép lại. Cột age thiếu có thể do thất lạc thông tin cá nhân của hành khách trong vụ tai nạn.
Đã trả lời cho câu hỏi cột nào thiếu dữ liệu nhiều nhất ( câu 7)

In [39]:
for col in ['sex', 'pclass', 'embarked']:
    print(f"--- Phân bổ cột {col} ---")
    print(df[col].value_counts())
    print("\n")

--- Phân bổ cột sex ---
sex
male      577
female    314
Name: count, dtype: int64


--- Phân bổ cột pclass ---
pclass
3    491
1    216
2    184
Name: count, dtype: int64


--- Phân bổ cột embarked ---
embarked
S    644
C    168
Q     77
Name: count, dtype: int64




* Sex: Số lượng nam giới chiếm ưu thế rõ rệt so với nữ giới trên tàu.
* Pclass: Hành khách hạng 3 (Hạng phổ thông) đông nhất, cho thấy Titanic chủ yếu phục vụ người lao động di cư bên cạnh các giới thượng lưu.
* Embarked: Đa số hành khách lên tàu tại cảng S (Southampton), đây có khả năng là cảng khởi hành chính của tàu.

In [40]:
# Phân tích 3 cột age, fare, sibsp
numeric_cols = ['age', 'fare', 'sibsp']
stats = df[numeric_cols].describe().loc[['mean', '50%', 'min', 'max']]
stats.rename(index={'50%': 'median'}, inplace=True)
print(stats)

              age        fare     sibsp
mean    29.699118   32.204208  0.523008
median  28.000000   14.454200  0.000000
min      0.420000    0.000000  0.000000
max     80.000000  512.329200  8.000000


* Fare : Có điểm bất thường cực lớn khi giá vé cao nhất (max) lên tới hơn 512, trong khi trung bình (mean) chỉ khoảng 32. Điều này chứng tỏ có một nhóm nhỏ khách siêu giàu kéo lệch biểu đồ.
* Age : Độ tuổi nhỏ nhất là trẻ sơ sinh (dưới 1 tuổi) và lớn nhất là 80, cho thấy đối tượng hành khách rất đa dạng.
* SibSp: Phần lớn mọi người đi một mình hoặc đi cùng 1 người thân (median = 0), nhưng có trường hợp đi cùng đến 8 người (anh chị em/vợ chồng), đây là một con số cần lưu ý khi phân tích nhóm gia đình.

# PHẦN C – PHÂN TÍCH CƠ BẢN

In [41]:
# Tính tỉ lệ sống sót (giá trị trung bình của cột survived)
overall_survival_rate = df['survived'].mean()
print(f"Tỉ lệ sống sót toàn tàu: {overall_survival_rate:.2%}")

Tỉ lệ sống sót toàn tàu: 38.38%


* Con số này cho biết trong tổng số hành khách trên tàu, chỉ có khoảng 38.38% may mắn sống sót. Điều này phản ánh tính chất khốc liệt của thảm họa Titanic khi hơn 60% hành khách đã không thể vượt qua.

In [42]:
# Tỉ lệ sống sót theo giới tính
print(df.groupby('sex')['survived'].mean())

# Tỉ lệ sống sót theo hạng vé
print(df.groupby('pclass')['survived'].mean())

sex
female    0.742038
male      0.188908
Name: survived, dtype: float64
pclass
1    0.629630
2    0.472826
3    0.242363
Name: survived, dtype: float64


* Giới tính: Có sự chênh lệch cực lớn khi tỉ lệ sống sót của Nữ giới cao áp đảo Nam giới. Điều này minh chứng cho quy tắc "Phụ nữ và trẻ em được ưu tiên".

* Hạng vé: Hành khách khoang hạng nhất (pclass 1) có cơ hội sống sót cao nhất (trên 60%), trong khi hạng 3 chỉ khoảng 24%. Địa vị kinh tế đóng vai trò quan trọng trong việc tiếp cận các tàu cứu hộ.
* Đã trả lời câu hỏi 1 và 2 và 5

# Trả lời 3 câu hỏi
3. Tỉ lệ trẻ em sống sót? ( Tôi dự đoán kết quả sẽ là <20%)
4. Những người đi cùng anh chị em hoặc vợ/chồng có tỉ lệ sống sót cao hơn người đi một mình không? (Dự đoán: Có, vì họ hỗ trợ nhau)
8. Có phải người càng lớn tuổi tỉ lệ sống sót càng thấp? ( Dự đoán : đúng, vì không đủ sức chịu đựng)

In [43]:
# Tạo cột tạm thời để kiểm tra đi một mình hay có người thân đi cùng
print(df.groupby('alone')['survived'].mean())

alone
False    0.505650
True     0.303538
Name: survived, dtype: float64


* Kết quả: Những người có người thân đi cùng (is_alone = False) có tỉ lệ sống cao hơn người đi một mình.
* Đối chiếu: ĐÚNG so với dự đoán ban đầu.
* Nhận xét: Việc có người thân đi cùng có thể giúp hành khách hỗ trợ nhau tốt hơn hoặc họ thuộc nhóm gia đình được ưu tiên lên thuyền.

In [44]:
# Tạo cột tạm thời để phân loại trẻ em
df['is_child'] = df['age'] < 10
survival_child = df.groupby('is_child')['survived'].mean()
print(f"Tỉ lệ sống sót của trẻ em: {survival_child[True]:.2%}")

Tỉ lệ sống sót của trẻ em: 61.29%


* Giải thích: Kết quả cho thấy tỉ lệ sống sót của trẻ em cao hơn hẳn mức trung bình toàn tàu.  
* Đối chiếu dự đoán: SAI. Tôi dự đoán dưới 20%, nhưng thực tế cao hơn rất nhiều.  
* Nhận xét: Thật bất ngờ vì trẻ em thường sẽ khó có thể sinh tồn trong thời tiết khắt nghiệt như vậy.

In [45]:
# Chia độ tuổi thành các nhóm 0-18, 18-30, 30-50, 50-80
age_bins = [0, 18, 30, 50, 80]
df['age_group'] = pd.cut(df['age'], bins=age_bins)
print(df.groupby('age_group', observed=True)['survived'].mean())

age_group
(0, 18]     0.503597
(18, 30]    0.355556
(30, 50]    0.423237
(50, 80]    0.343750
Name: survived, dtype: float64


* Phân tích: Tỉ lệ sống sót cao nhất nằm ở nhóm nhỏ tuổi nhất và giảm dần ở các nhóm tuổi lao động, nhưng nhóm người rất già đôi khi lại có tỉ lệ biến động.  
* Đối chiếu dự đoán: ĐÚNG MỘT PHẦN.
* Nhận xét: Mặc dù sức chịu đựng của người già thấp, nhưng yếu tố quan trọng nhất vẫn là sự ưu tiên. Nhóm trẻ tuổi luôn có tỉ lệ sống cao nhất, trong khi các nhóm tuổi trung niên trở lên có tỉ lệ thấp hơn rõ rệt.

### Nhận xét:
* Về tỉ lệ sống sót của trẻ em: Thực tế cho thấy tỉ lệ sống sót của trẻ em là rất cao (thường trên 50%), trái ngược hoàn toàn với dự đoán ban đầu của tôi là dưới 20%. Điều này chứng tỏ trong tình huống khẩn cấp, quy tắc đạo đức "ưu tiên trẻ em" đã được thực thi hiệu quả hơn so với trực giác ban đầu về sức chịu đựng thể chất kém của các em.
* Về việc đi cùng người thân (SibSp): Kết quả phân tích ủng hộ dự đoán ban đầu của tôi. Những người đi cùng anh chị em hoặc vợ/chồng có tỉ lệ sống sót cao hơn so với những người đi đơn lẻ. Sự hỗ trợ lẫn nhau giữa các thành viên trong gia đình thực sự là một yếu tố tích cực giúp tăng khả năng sống sót.
* Về mối liên hệ giữa độ tuổi và tỉ lệ sống: Dự đoán của tôi là đúng một phần. Nhóm trẻ em luôn có tỉ lệ sống cao nhất, trong khi các nhóm tuổi lớn hơn có tỉ lệ sống thấp hơn rõ rệt. Tuy nhiên, dữ liệu cũng cho thấy sự ưu tiên về giới tính và hạng vé đôi khi còn quan trọng hơn cả yếu tố sức khỏe do tuổi tác.

# PHẦN D – CÂU HỎI MỚI (SAU PHÂN TÍCH)
1. Liệu một người đàn ông ở khoang hạng nhất có tỉ lệ sống sót cao hơn một người phụ nữ ở khoang hạng ba không? 
2. Những gia đình đi cực kỳ đông người (>5) có tỉ lệ sống thấp hơn hẳn các gia đình nhỏ không?
3. Khoang nào thì tỉ lệ sống sót sẽ ít chênh lệch giữa hai giới?
4. Tỉ lệ sống sót của trẻ em ở 3 khoang?
5. Có trẻ em đi một mình không? 

In [46]:
# Dùng pivot table để so sánh chéo
comparison = df.pivot_table('survived', index='sex', columns='pclass')
print(comparison)

pclass         1         2         3
sex                                 
female  0.968085  0.921053  0.500000
male    0.368852  0.157407  0.135447


* Phụ nữ ở khoang hạng 3 vẫn có tỉ lệ sống (~50%) cao hơn Đàn ông khoang hạng 1 (~36%).
* Thật sự bất ngờ khi tỉ lệ sống sót của của hai giới lại chênh lệch nặng tới như vậy. Lúc này thì tiền không mua được mạng sống rồi.


In [47]:
# Những gia đình đi cực kỳ đông người >5 có tỉ lệ sống thấp hơn hẳn các gia đình nhỏ không?
df['family_size'] = df['sibsp'] + df['parch']
print(df.groupby('family_size')['survived'].mean())

family_size
0     0.303538
1     0.552795
2     0.578431
3     0.724138
4     0.200000
5     0.136364
6     0.333333
7     0.000000
10    0.000000
Name: survived, dtype: float64


* Ta thấy tỉ lệ sống sót tăng lên ở nhóm gia đình nhỏ (1-3 người) nhưng lại giảm thê thảm ở các đại gia đình từ 4 người trở lên. Có lẽ vì quá đông nên họ khó tìm thấy nhau hoặc không đủ chỗ trên cùng một thuyền cứu hộ.